### PR TITLE
feature/salesforce-marketing-cloud

### PR DESCRIPTION
**Pull request by mmsarom**
Adding a new Provider & Client for Salesforce Marketing Cloud (SFMC).

Following similar pattern to existing Salesforce Sales/Services Cloud provider by utilising existing API User's credentials (has to be manually created in SFMC) to rotate its own secret. **Zeep** package added to assist with constructing and making SOAP calls.

**Pre-requisite prior to use:**
- API User setup in SFMC
- API User provided with permissions (Admin -> API Access)

Ps. Has been a while since I used python. Let me know if anything's butchered real bad 😅

### CODE REVIEW

File: src/keydra/providers/salesforce_marketing_cloud.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+from keydra.exceptions import ConfigException, DistributionException
+from keydra.exceptions import RotationException
+
+from keydra.logging import get_logger
+
+LOGGER = get_logger()
+
+PASS_LENGTH = 32
+
+
```
Comment (mmsarom) [inline src/keydra/providers/salesforce_marketing_cloud.py:21]: Keen to get some thoughts on this. It's feasible to combine the new provider with the existing Salesforce provider. I opted against it to keep it more maintainable if Salesforce Sales Cloud differs greately from Salesforce Marketing Cloud in the long run.

File: src/keydra/clients/salesforce_marketing_cloud.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        user = account_user_obj(
+            UserID=username,
+            Name=username,
+            Email=username,
+            Password=newpassword,
+            BusinessUnit=self._businessUnit,
+            Client=sfmc_client_obj(ID=self._mid)
+        )
+
+        # Check Sever is OK before posting a User Update
```
Comment (gyrospectre) [inline src/keydra/clients/salesforce_marketing_cloud.py:85]: Typo?

File: src/keydra/providers/salesforce_marketing_cloud.py (hunk 1 above)
Comment (gyrospectre) [reply src/keydra/providers/salesforce_marketing_cloud.py:21]: I agree tbh. I think for now, separate makes more sense. I get the felling that SFMC is somewhat separate from core SF in terms of platform, so feel the providers should be separate too. I assume that you did not find you had to duplicate large chunks of code? 

We've been here before with the Splunk and Splunk HEC providers - we opted for separate there as well after some debate.

**Comment by gyrospectre** [comment]
Looks pretty good @mmsarom! Just one typo which is not a blocker, and just calling out you've neglected to add any doco for the new provider.

I'll approve/merge this now, maybe once you have it working in your environment you can circle back and add the docs?

Thanks for your contribution!

**Review by gyrospectre — APPROVED** [review]
Looks pretty good @mmsarom! Just one typo which is not a blocker, and just calling out you've neglected to add any doco for the new provider.

I'll approve/merge this now, maybe once you have it working in your environment you can circle back and add the docs?

Thanks for your contribution!

File: src/keydra/providers/salesforce_marketing_cloud.py (hunk 1 above)
Comment (mmsarom) [reply src/keydra/providers/salesforce_marketing_cloud.py:21]: 👍  Duplicated the structure, but did change a few things given the Client for SFMC is vastly different.

**Pull request merged**